src/Interpreters/MonotonousOrderByVisitor.h

```diff
@@ -0,0 +1,194 @@
+        ASTPtr & identifier;
+        bool & is_monotonous;
+        bool & is_positive;
+        DataTypePtr data_type = nullptr;
+        bool first = true;
+        bool done = false;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (!data.done && ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (!data.is_monotonous)
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
+        {
+            data.is_monotonous = false;
+            data.done = true;
+            return;
+        }
+
+        if (data.first)
+            data.identifier = ast_function.arguments->children[0];
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function.name, data.context);
+        auto pos = IdentifierSemantic::chooseTable(*data.identifier->as<ASTIdentifier>(), data.tables, true);
+        auto data_type_and_name = data.tables[*pos].columns.tryGetByName(data.identifier->getAliasOrColumnName());
+        if (data.first)
+        {
+            data.data_type = data_type_and_name->type;
+            data.first = false;
+        }
+        auto name = data_type_and_name->name;
+
+        ColumnsWithTypeAndName args;
+        args.emplace_back(data.data_type, name);
+        auto function_base = function->build(args);
+
+        auto cur_data = data.data_type;
+        data.data_type = function_base->getReturnType();
+
+        if (!function_base->hasInformationAboutMonotonicity())
+        {
+            data.is_monotonous = false;
+            data.done = true;
+            return;
+        }
+        auto monotonicity = function_base->getMonotonicityForRange(*cur_data, NULL, NULL);
+        if (!monotonicity.is_always_monotonic)
+        {
+            data.is_monotonous = false;
+            data.done = true;
+            return;
+        }
+
+        if (!monotonicity.is_positive)
+            data.is_positive = !data.is_positive;
+    }
+
+    static bool needChildVisit(const ASTPtr &, const ASTPtr &)
+    {
+        return true;
+    }
+
+};
+
+using MonotonicityCheckVisitor = InDepthNodeVisitor<MonotonicityCheckMatcher, false>;
+
+class FirstLevelASTFunctionMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        ASTPtr & identifier;
+        bool & is_positive;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (ast_function.children.size() != 1)
+            return;
+
+        bool is_monotonous = true;
```

Comment:
It's better to use `Monotonicity` struct as well instead of two splitted bool values. Even if we do not need some of them now.